src/data_gradients/dataset_adapters/formatters/segmentation.py

```diff
@@ -56,11 +60,11 @@ def format(self, images: Tensor, labels: Tensor) -> Tuple[Tensor, Tensor]:
 
         images = check_images_shape(images, n_image_channels=self.get_n_image_channels(images=images))
 
-        labels = self.validate_labels_dim(labels, n_classes=self.data_config.n_classes, ignore_labels=self.ignore_labels)
-        labels = self.ensure_hard_labels(labels, n_classes=self.data_config.n_classes, threshold_value=self.threshold_value)
+        labels = self.validate_labels_dim(labels, n_classes=self.data_config.get_n_classes(), ignore_labels=self.ignore_labels)
```

Comment:
you don't really need to change this.
just use @property instead

Reply:
Making `data_config .n_classes` a property?
I think it's a good idea, the issue is that `n_classes` is already an attribute. It basically represents the cached value so it may be None if it was not initialized/asked to the user.

I think we should change the cache attributes of `data_config` to be private; `_n_classes` would represent the cache, while the property `n_classes` would be the interface for the user, and would call `get_n_classes()` under the hood.
I didn't do it before because it felt overengineering back in the time, but now that we are starting to have many attributes I think we are really starting to feel the need.

Reply:
If we use Pydantic.BaseModel, we can use `alias`. This enables the instantiation of a variable using its alias:

```python

class MyClass(BaseModel):
    num_classes_: List[str] = Field(alias="num_classes")

    @property
    def num_classes(self) -> List[str]:
        if self.num_classes_ is None:
            self.num_classes_ = 4  # Ask question here instead of dummy code
        return self.num_classes_

obj = MyClass(num_classes=4)
print(obj.num_classes)
```

Without the `alias`, we need would need to isntantiate the object like this:
```python
obj = MyClass(_num_classes=4)
print(obj.num_classes)
```
Which is very confusing for the user

Reply:
Just not sure if it's not overengineered. This is why until now I prefered to use the following logic
```python

class MyClass(BaseModel):
    num_classes: List[str]

    def get_num_classes(self) -> List[str]:
        if self.num_classes_is None:
            self.num_classes = 4  # Ask question here instead of dummy code
        return self.num_classes

obj = MyClass(num_classes=4)
print(obj.get_num_classes())
```

Reply:
@ofrimasad update
On thursday we talked about using properties "setter", but it doesnt work

```python
class MyClass(BaseModel):
    _num_classes: List[str] = []

    @property
    def num_classes(self) -> List[str]:
        if self._num_classes is None:
            self._num_classes = 4  # Ask question here instead of dummy code
        return self._num_classes

    @num_classes.setter
    def num_classes(self, value):
        self._num_classes = value

obj = MyClass(num_classes=4) # Does not work, the setter is not recognized as `__init__` param
```

So, I think let's keep the code as is for now to not lose our focus. Then, we can think about a way to make it cleaner in the future.

Reply:
ok. so lets keep it this way